codec/src/main/java/io/netty/handler/codec/base64/Base64.java

```diff
@@ -230,86 +293,114 @@ public static ByteBuf decode(
             throw new NullPointerException("dialect");
         }
 
-        byte[] DECODABET = decodabet(dialect);
+        // Using a ByteProcessor to reduce bound and reference count checking.
+        return new Decoder().decode(src, off, len, allocator, dialect);
+    }
+
+    private static final class Decoder implements ByteProcessor {
+
+        private final byte[] b4 = new byte[4];
+        private int b4Posn;
+        private byte sbiCrop;
+        private byte sbiDecode;
+        private byte[] decodabet;
+        private int outBuffPosn;
+        private ByteBuf dest;
 
-        int len34 = len * 3 / 4;
-        ByteBuf dest = allocator.buffer(len34).order(src.order()); // Upper limit on size of output
-        int outBuffPosn = 0;
+        ByteBuf decode(ByteBuf src, int off, int len, ByteBufAllocator allocator, Base64Dialect dialect) {
+            int len34 = len * 3 / 4;
+            dest = allocator.buffer(len34).order(src.order()); // Upper limit on size of output
+
+            decodabet = decodabet(dialect);
+            boolean release = true;
+            try {
+                src.forEachByte(off, len, this);
+                ByteBuf slice = dest.slice(0, outBuffPosn);
+                release = false;
+                return slice;
+            } finally {
+                if (release) {
```

Comment:
nit: to avoid an additional variable you can simplify to:

```java
            try {
                src.forEachByte(off, len, this);
                return dest.slice(0, outBuffPosn);
            } catch (Throwable cause) {
                dest.release();
                PlatformDependent.throwException(cause);
                return null;
            }
```